### PR TITLE
CustomCluster configuration memory optimization

### PR DESCRIPTION
**Pull request by DiazAlan**
Commit Message: Saves memory in cluster info
Additional Description: In cases where CustomClusterType is not used, the unique_ptr will be null and save memory
Risk Level: Low
Testing: Changed mock for new typing
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: envoy/upstream/upstream.h
Comment (adisuissa) [inline envoy/upstream/upstream.h:971]: nit: return a const CustomClusterType

File: source/common/upstream/upstream_impl.h
Comment (adisuissa) [inline source/common/upstream/upstream_impl.h:765]: nit: return const CustomClusterType.

**Review by adisuissa — COMMENTED** [review]
LGTM modulo nit.
Thanks!

File: source/common/upstream/upstream_impl.h
Comment (DiazAlan) [reply source/common/upstream/upstream_impl.h]: Done

File: envoy/upstream/upstream.h
Comment (DiazAlan) [reply envoy/upstream/upstream.h]: Done

**Comment by DiazAlan** [comment]
@yurykats @adisuissa

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

**Comment by adisuissa** [comment]
CI failure seems relevant to the PR changes. PTAL.

**Review by zuercher — DISMISSED** [review]

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc
Comment (adisuissa) [inline test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc:1041]: nit: Can this directly return `makeOptRef(cluster_type_)` (avoiding the temp var above)?

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc
Comment (adisuissa) [inline test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc:1063]: Same as above

**Review by adisuissa — COMMENTED** [review]
Thanks, overall LGTM.
Left one minor comment/question.

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc
Comment (DiazAlan) [reply test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc]: Done

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc
Comment (DiazAlan) [reply test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc]: Done

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
/assign-from @envoyproxy/senior-maintainers

**Review by zuercher — APPROVED** [review]

**Pull request merged**
